libp2p/src/builder.rs

```diff
@@ -0,0 +1,399 @@
+use libp2p_core::{muxing::StreamMuxerBox, Transport};
+use libp2p_swarm::NetworkBehaviour;
+use std::marker::PhantomData;
+
+pub struct SwarmBuilder {}
+
+impl SwarmBuilder {
+    pub fn new() -> SwarmBuilder {
+        Self {}
+    }
+
+    pub fn with_new_identity(self) -> ProviderBuilder {
```

Comment:
Makes sense. That said, I don't know how. I could introduce a wrapper `struct` that delegates to a generic inner type. Though the delegation logic would introduce a lot of complexity.

Reply:
It does not necessarily reduce the number of types but what I've seen in other projects:

- One central `Builder` struct that has a generic state or phase
- The phases have fields themselves

See `rustls` for example: https://github.com/rustls/rustls/blob/main/rustls/src/builder.rs#L97

Something similar for us would be:

- `SwarmBuilder`
- states / phases called:
   - `WantsBaseTransport`
   - `WantsRelay`
   - `WantsOtherTransports`
- `impl` blocks for `SwarmBuilder<WantsRelay>` etc

Reply:
Problem is that the "phases" differ in the number of type parameters they require:

``` rust
pub struct SwarmBuilder {}
pub struct ProviderBuilder {
pub struct TcpBuilder<P> {
pub struct RelayBuilder<P, T> {
pub struct OtherTransportBuilder<P, T, R> {
pub struct DnsBuilder<P, T, R> {
pub struct BehaviourBuilder<P, R> {
pub struct Builder<P, B> {
```

We could parameterize `SwarmBuilder` over the phase + all type parameters across all phases:

``` rust
pub struct SwarmBuilder<Phase, Provider, Transport, Relay, Behaviour>
```

Reply:
If the `SwarmBuilder` has a single type parameter for the phase, this should not matter? For example, `WantsBaseTransport` can have another type parameter for being generic over the supported runtime which can be passed on from a `WantsRuntime` phase, right?

Reply:
If `WantsBaseTransport` is generic over a type parameter, then `SwarmBuilder` needs to provide that type parameter to `WantsBaseTransport` and thus needs to be generic over that same type parameter as well. Thereby I am not sure how your suggestion simplifies things. @thomaseizinger can you expand?

Reply:
> Maybe I am missing something but I thought that this could work.

Exactly the opposite. You are right. It is possible. Thank you for being meticulous here @thomaseizinger.



> It does not necessarily reduce the number of types

This still stands. But I think having a single root level type is easier for users to handle. In other words, I am in favor of the phase-style.

See d3ffdd4d2